File: core/events.py
"""
core/events.py

JSONL event types emitted by the Python backend to the CLI.
These mirror src/backend/events.js so the TUI parser can consume them.
"""

from __future__ import annotations

import json
import time
from dataclasses import dataclass, field
from typing import Any, Optional


# ── Event type constants (kept in sync with suryafool-cli/src/backend/events.js) ──

COMMAND_STARTED = "command.started"
COMMAND_OUTPUT = "command.output"
COMMAND_PROGRESS = "command.progress"
COMMAND_COMPLETED = "command.completed"
COMMAND_FAILED = "command.failed"

FINDING_CREATED = "finding.created"
SCAN_STARTED = "scan.started"
SCAN_PROGRESS = "scan.progress"
SCAN_COMPLETED = "scan.completed"

AGENT_STATUS = "agent.status"
AGENT_STARTED = "agent.started"
AGENT_STOPPED = "agent.stopped"

VULN_FOUND = "vuln.found"

# Phase 2.7.5 — durable evidence record produced by a capture-style
# capability. Mirror in suryafool-cli/src/backend/events.js.
EVIDENCE_CREATED = "evidence.created"

LOG = "log"
ERROR = "error"

LLM_RESPONSE = "llm.response"


@dataclass
class Event:
    type: str
    timestamp: float = field(default_factory=time.time)
    data: dict[str, Any] = field(default_factory=dict)

    def to_jsonl(self) -> str:
        payload = {"type": self.type, "timestamp": self.timestamp, **self.data}
        return json.dumps(payload, separators=(",", ":"), default=str)


# ── Helpers ────────────────────────────────────────────────────────────────────

def event(type_: str, **data: Any) -> Event:
    return Event(type=type_, data=data)


def emit(stream, type_: str, **data: Any) -> None:
    """Emit a JSONL event to a writable stream. Safe no-op if stream is None."""
    if stream is None:
        return
    e = event(type_, **data)
    stream.write(e.to_jsonl() + "\n")
    stream.flush()

File: suryafool-cli/src/backend/events.js
// src/backend/events.js
// Structured JSONL event protocol for backend communication

/**
 * Event types emitted by the backend
 * Each event is a JSON object on its own line (JSONL format)
 */

export const EventType = {
  // Command lifecycle
  COMMAND_STARTED: 'command.started',
  COMMAND_OUTPUT: 'command.output',
  COMMAND_PROGRESS: 'command.progress',
  COMMAND_COMPLETED: 'command.completed',
  COMMAND_FAILED: 'command.failed',
  
  // Findings & scanning
  FINDING_CREATED: 'finding.created',
  SCAN_STARTED: 'scan.started',
  SCAN_PROGRESS: 'scan.progress',
  SCAN_COMPLETED: 'scan.completed',
  
  // Agent status
  AGENT_STATUS: 'agent.status',
  AGENT_STARTED: 'agent.started',
  AGENT_STOPPED: 'agent.stopped',
  
  // Vulnerabilities
  VULN_FOUND: 'vuln.found',

  // Evidence (Phase 2.7.5+ — durable capture records)
  EVIDENCE_CREATED: 'evidence.created',

  // Logs
  LOG: 'log',
  ERROR: 'error',
  
  // LLM
  LLM_RESPONSE: 'llm.response',
};

/**
 * Create a command.started event
 */
export function commandStarted(command, args) {
  return { type: EventType.COMMAND_STARTED, command, args, timestamp: Date.now() };
}

/**
 * Create a command.output event
 */
export function commandOutput(line) {
  return { type: EventType.COMMAND_OUTPUT, line, timestamp: Date.now() };
}

/**
 * Create a command.progress event
 */
export function commandProgress(progress, phase, message) {
  return { type: EventType.COMMAND_PROGRESS, progress, phase, message, timestamp: Date.now() };
}

/**
 * Create a command.completed event
 */
export function commandCompleted(command, result) {
  return { type: EventType.COMMAND_COMPLETED, command, result, timestamp: Date.now() };
}

/**
 * Create a command.failed event
 */
export function commandFailed(command, error) {
  return { type: EventType.COMMAND_FAILED, command, error, timestamp: Date.now() };
}

/**
 * Create a finding.created event
 */
export function findingCreated(finding) {
  return { type: EventType.FINDING_CREATED, finding, timestamp: Date.now() };
}

/**
 * Create a scan.progress event
 */
export function scanProgress(progress, phase, message) {
  return { type: EventType.SCAN_PROGRESS, progress, phase, message, timestamp: Date.now() };
}

/**
 * Create an agent.status event
 */
export function agentStatus(agent, status, metadata) {
  return { type: EventType.AGENT_STATUS, agent, status, metadata, timestamp: Date.now() };
}

/**
 * Create a vuln.found event
 */
export function vulnFound(vulnerability) {
  return { type: EventType.VULN_FOUND, vulnerability, timestamp: Date.now() };
}

/**
 * Create an evidence.created event (Phase 2.7.5+)
 */
export function evidenceCreated(evidence, sourceActionId, runId) {
  return { type: EventType.EVIDENCE_CREATED, evidence, sourceActionId, runId, timestamp: Date.now() };
}

/**
 * Create a log event
 */
export function logEvent(level, message, metadata) {
  return { type: EventType.LOG, level, message, metadata, timestamp: Date.now() };
}

/**
 * Create an error event
 */
export function errorEvent(message, details) {
  return { type: EventType.ERROR, message, details, timestamp: Date.now() };
}

/**
 * Validate if an object is a valid event
 */
export function isValidEvent(obj) {
  return obj && typeof obj === 'object' && typeof obj.type === 'string' && Object.values(EventType).includes(obj.type);
}

export default EventType;